### PR TITLE
Autoencoder fixes

### PR DESCRIPTION
**Pull request by epierson9**
@kohpangwei small PR PTAL if you have a chance. code runs.

### CODE REVIEW

File: general_autoencoder.py [hunk 1]
```diff
@@ -142,6 +143,12 @@ def decode(self, Z):
     def get_loss():
         raise NotImplementedError
 
+    def get_longitudinal_loss():
+        raise NotImplementedError
+
+    def get_Z1_from_Z0(self, Z0):
```
Comment (epierson9) [inline general_autoencoder.py]: we have a fast_forward_Z method -- how will this interact with it, and are we introducing redundancy?

File: general_autoencoder.py [hunk 2]
```diff
@@ -316,11 +327,31 @@ def _fit_from_processed_data(self,
                 self.Z = self.encode(self.X)
                 
             self.Xr = self.decode(self.Z)
-            self.combined_loss, self.binary_loss, self.continuous_loss, self.reg_loss = self.get_loss()
+            self.combined_loss, self.binary_loss, self.continuous_loss, self.reg_loss = self.get_loss(self.X, self.Xr)
```
Comment (epierson9) [inline general_autoencoder.py]: are self.X and self.Xr sufficient to compute the loss in all cases? won't we also need, eg, self.ages in some cases?

File: general_autoencoder.py [hunk 3]
```diff
@@ -316,11 +327,31 @@ def _fit_from_processed_data(self,
...
             self.optimizer = self.optimization_method(learning_rate=self.learning_rate).minimize(self.combined_loss)
+
             if self.uses_longitudinal_data:
-                self.combined_longitudinal_loss, self.binary_longitudinal_loss, self.continuous_longitudinal_loss, self.reg_longitudinal_loss = self.get_longitudinal_loss()
-                self.longitudinal_optimizer = self.optimization_method(learning_rate=self.learning_rate).minimize(self.combined_longitudinal_loss)
+
+                # now project Z0 forward to get Z1. 
+                # This requires multiplying the age components by longitudinal_ages1 / ages
+                self.Z1 = self.get_Z1_from_Z0(self.Z)
```
Comment (epierson9) [inline general_autoencoder.py]: is this call signature sufficient? to get Z1, we need to know the original age + the longitudinal age.

File: general_autoencoder.py [hunk 4]
```diff
@@ -340,24 +371,29 @@ def _fit_from_processed_data(self,
...
+                regularization_weighting_for_epoch = self.get_regularization_weighting_for_epoch(
+                    epoch,
+                    verbose=verbose)
  
                 self._train_epoch(regularization_weighting_for_epoch)
                     
                 if (epoch % self.num_epochs_before_eval == 0) or (epoch == self.max_epochs - 1):
                     
+                    # regularization weighting is set to 1 just for the purpose of printing out 
+                    # the losses
```
Comment (epierson9) [inline general_autoencoder.py]: i think this will also affect the saved losses (ie, we save a record of losses, and those will now be computed with regularization weighting = 1) which is fine, but something to be aware of. i don't think we would expect the train_mean_combined_loss to decrease monotonically, since that's not what we're actually optimizing.

File: laplacian_variational_autoencoder.py [hunk 5]
```diff
@@ -55,11 +55,11 @@ def encode(self, X):
         return Z
 
 
-    def get_loss(self):
+    def get_loss(self, X, Xr):
```
Comment (epierson9) [inline laplacian_variational_autoencoder.py]: (i am assuming you just went through all the autoencoders + changed any call signatures on get_loss + removed references to self.X and self.Xr?)

File: sparse_correlation_variational_age_autoencoder.py [hunk 6]
```diff
@@ -70,21 +70,23 @@ def compute_correlation_sparsity_loss(self, Z, X):
         sparsity_loss = tf.reduce_sum(clipped_pearsonr_matrix)
         return sparsity_loss
     
-    def get_loss(self):
+    def get_loss(self, X, Xr):
         """
-        Adds a correlation sparsity loss to the regularization term. 
+        Adds a correlation sparsity loss to the regularization term.
+        TODO: The use of self.age_adjusted_X here is pretty bad code-design-wise,
+        since we pass in X as an argument. Revisit when we start using this model again.
```
Comment (epierson9) [inline sparse_correlation_variational_age_autoencoder.py]: there are references to other fields as self as well, though -- eg, self.Z.

File: variational_longitudinal_rate_of_aging_autoencoder.py [hunk 7]
```diff
@@ -164,51 +164,32 @@ def fill_feed_dict_longitudinal(self,
...
-        binary_loss_0 = self.get_binary_loss(X0_binary, Xr0_logits)
-        continuous_loss_0 = self.get_continuous_loss(X0_continuous, Xr0_continuous)
-        binary_loss_1 = self.get_binary_loss(X1_binary, Xr1_logits)
-        continuous_loss_1 = self.get_continuous_loss(X1_continuous, Xr1_continuous)
+    def get_longitudinal_loss(self, 
+                              binary_loss0, 
+                              continuous_loss0, 
+                              binary_loss1,
+                              continuous_loss1,
+                              reg_loss):
```
Comment (epierson9) [inline variational_longitudinal_rate_of_aging_autoencoder.py]: do you think the variable names in variational_longitudinal_rate_of_aging_autoencoder make sense? (longitudinal_X1 etc.) i was worried when i chose them they were confusing.

File: general_autoencoder.py [hunk 8]
```diff
@@ -316,11 +327,31 @@ def _fit_from_processed_data(self,
...
+                # now project Z0 forward to get Z1. 
+                # This requires multiplying the age components by longitudinal_ages1 / ages
+                self.Z1 = self.get_Z1_from_Z0(self.Z)
+
+                # reconstruct X1 from Z1.
+                self.Xr1 = self.decode(self.Z1)
+
+                _, self.binary_loss1, self.continuous_loss1, _ = self.get_loss(self.longitudinal_X1, self.Xr1)
+
+                self.combined_longitudinal_loss, self.binary_longitudinal_loss, \
```
Comment (epierson9) [inline general_autoencoder.py]: given that what's being done by get_longitudinal_loss is very simple (we just add them together and reweight) i wonder if it would be clearer to pull remove self.get_longitudinal_loss or pull it out of the subclass. maybe not.

File: laplacian_variational_autoencoder.py (hunk 5 above)
Comment (kohpangwei) [reply laplacian_variational_autoencoder.py]: Yup

File: general_autoencoder.py (hunk 1 above)
Comment (kohpangwei) [reply general_autoencoder.py]: Hm. I tried to reconcile them, and realized that it's a bit harder than I thought. `get_Z1_from_Z0` is defining a new tensor, whereas our `fast_forward_Z` method takes in actual numbers and adjusts them. I'll add a comment to this effect; perhaps it's possible to treat them more coherently but not a huge deal now I think.

File: standard_autoencoder.py [hunk 9]
```diff
@@ -27,6 +27,10 @@ def __init__(self,
         self.learn_continuous_variance = learn_continuous_variance
 
     def init_network(self):
+        if self.learn_continuous_variance:
+            # we exponentiate this because it has to be non-negative. 
+            self.log_continuous_variance = tf.Variable(self.initialization_function([1]))
```
Comment (kohpangwei) [inline standard_autoencoder.py]: I wonder if this should be on a per-phenotype basis?

File: general_autoencoder.py [hunk 10]
```diff
@@ -161,14 +165,16 @@ def get_ages(self, df):
         ages = np.array(df['age_sex___age'].values, dtype=np.float32)
         return self.age_preprocessing_function(ages)
             
-    def combine_loss_components(self, binary_loss, continuous_loss, regularization_loss):
-        return binary_loss + continuous_loss + self.regularization_weighting * regularization_loss
+    def combine_loss_components(self, binary_loss, continuous_loss, regularization_loss, regularization_weighting):
+        return binary_loss + continuous_loss + regularization_weighting * regularization_loss
```
[the remark this answers is not in the document]
Comment (epierson9) [reply general_autoencoder.py]: done.

File: general_autoencoder.py [hunk 11]
```diff
@@ -288,33 +309,57 @@ def _fit_from_processed_data(self,
...
+                self.lon_Xr0 = self.decode(self.lon_Z0)
+                self.lon_Xr1 = self.decode(self.lon_Z1)
+                self.lon_binary_loss0, self.lon_continuous_loss0 = self.get_binary_and_continuous_loss(self.lon_X0, self.lon_Xr0)
+                self.lon_binary_loss1, self.lon_continuous_loss1 = self.get_binary_and_continuous_loss(self.lon_X0, self.lon_Xr0)
+
+                # multiply all loss components by longitudinal loss weighting factor
+                binary_lon_loss = (self.lon_binary_loss0 + self.lon_binary_loss1) * self.lon_loss_weighting_factor
+                continuous_lon_loss = (self.lon_continuous_loss0 + self.lon_continuous_loss1) * self.lon_loss_weighting_factor
+                
+                reg_lon_loss = self.reg_lon_loss * self.lon_loss_weighting_factor
```
[the remark this answers is not in the document]
Comment (epierson9) [reply general_autoencoder.py]: i think so? good catch. this seems like potentially a significant change (worth rerunning models that didn't do this, i think?)

File: general_autoencoder.py [hunk 12]
```diff
@@ -288,33 +309,57 @@ def _fit_from_processed_data(self,
...
+                self.combined_lon_loss = binary_lon_loss + continuous_lon_loss + reg_lon_loss
+                
+                self.combined_cross_sectional_plus_lon_loss = (self.combined_lon_loss + self.combined_loss)
+                self.optimizer = self.optimization_method(learning_rate=self.learning_rate).minimize(
+                    self.combined_cross_sectional_plus_lon_loss)
```
[the remark this answers is not in the document]
Comment (epierson9) [reply general_autoencoder.py]: hm, fair point. moved this code to longitudinal autoencoder.

File: general_autoencoder.py [hunk 13]
```diff
@@ -140,8 +144,8 @@ def decode(self, Z):
         raise NotImplementedError
 
     def get_loss():
-        raise NotImplementedError
-
+        return self.combined_loss, self.binary_loss, self.continuous_loss, self.reg_loss
```
[the remark this answers is not in the document]
Comment (epierson9) [reply general_autoencoder.py]: naw, i'll cut.

File: standard_autoencoder.py (hunk 9 above)
Comment (epierson9) [reply standard_autoencoder.py]: as discussed, this corresponds to shifting from PPCA to factor analysis, i think? maybe worth doing at some point, but will leave as is for now.
